Consider skipping constructors for non-public destructor

`chimera` generates constructor bindings even when the destructor is not public, and this leads to compilation error something like this:
```cpp
In file included from /usr/include/dart/gui/osg/osg.hpp:8:0,
                 from /home/jslee02/dev/prl/dartpy/osg/src/generated/_ZN4dart3gui3osg9WorldNodeE.cpp:11:
/usr/include/c++/5/backward/auto_ptr.h: In instantiation of ‘std::auto_ptr< <template-parameter-1-1> >::~auto_ptr() [with _Tp = dart::gui::osg::WorldNode]’:
/usr/include/boost/python/make_constructor.hpp:48:35:   required from ‘void boost::python::detail::install_holder<T>::dispatch(U*, mpl_::true_) const [with U = dart::gui::osg::WorldNode; T = dart::gui::osg::WorldNode*; mpl_::true_ = mpl_::bool_<true>]’
/usr/include/boost/python/make_constructor.hpp:40:19:   required from ‘PyObject* boost::python::detail::install_holder<T>::operator()(T) const [with T = dart::gui::osg::WorldNode*; PyObject = _object]’
/usr/include/boost/python/detail/invoke.hpp:75:14:   required from ‘PyObject* boost::python::detail::invoke(boost::python::detail::invoke_tag_<false, false>, const RC&, F&, AC0&, AC1&) [with RC = boost::python::detail::install_holder<dart::gui::osg::WorldNode*>; F = _ZN4dart3gui3osg9WorldNodeE()::<lambda(std::shared_ptr<dart::simulation::World>, osg::ref_ptr<osgShadow::ShadowTechnique>)>; AC0 = boost::python::arg_from_python<std::shared_ptr<dart::simulation::World> >; AC1 = boost::python::arg_from_python<osg::ref_ptr<osgShadow::ShadowTechnique> >; PyObject = _object]’
/usr/include/boost/python/detail/caller.hpp:218:46:   required from ‘PyObject* boost::python::detail::caller_arity<2u>::impl<F, Policies, Sig>::operator()(PyObject*, PyObject*) [with F = _ZN4dart3gui3osg9WorldNodeE()::<lambda(std::shared_ptr<dart::simulation::World>, osg::ref_ptr<osgShadow::ShadowTechnique>)>; Policies = boost::python::detail::constructor_policy<boost::python::default_call_policies>; Sig = boost::mpl::v_item<dart::gui::osg::WorldNode*, boost::mpl::v_item<std::shared_ptr<dart::simulation::World>, boost::mpl::v_item<osg::ref_ptr<osgShadow::ShadowTechnique>, boost::mpl::vector0<mpl_::na>, 1>, 1>, 1>; PyObject = _object]’
/usr/include/boost/python/object/py_function.hpp:64:33:   required from ‘PyObject* boost::python::objects::signature_py_function_impl<Caller, Sig>::operator()(PyObject*, PyObject*) [with Caller = boost::python::detail::caller<_ZN4dart3gui3osg9WorldNodeE()::<lambda(std::shared_ptr<dart::simulation::World>, osg::ref_ptr<osgShadow::ShadowTechnique>)>, boost::python::detail::constructor_policy<boost::python::default_call_policies>, boost::mpl::v_item<dart::gui::osg::WorldNode*, boost::mpl::v_item<std::shared_ptr<dart::simulation::World>, boost::mpl::v_item<osg::ref_ptr<osgShadow::ShadowTechnique>, boost::mpl::vector0<mpl_::na>, 1>, 1>, 1> >; Sig = boost::mpl::v_item<void, boost::mpl::v_item<boost::python::api::object, boost::mpl::v_mask<boost::mpl::v_item<dart::gui::osg::WorldNode*, boost::mpl::v_item<std::shared_ptr<dart::simulation::World>, boost::mpl::v_item<osg::ref_ptr<osgShadow::ShadowTechnique>, boost::mpl::vector0<mpl_::na>, 1>, 1>, 1>, 1>, 1>, 1>; PyObject = _object]’
/home/jslee02/dev/prl/dartpy/osg/src/generated/_ZN4dart3gui3osg9WorldNodeE.cpp:50:1:   required from here
/usr/include/dart/gui/osg/WorldNode.hpp:152:11: error: ‘virtual dart::gui::osg::WorldNode::~WorldNode()’ is protected
   virtual ~WorldNode();
           ^
In file included from /usr/include/c++/5/memory:85:0,
                 from /home/jslee02/dev/prl/dartpy/osg/include/dartpy/pointers.h:3,
                 from /home/jslee02/dev/prl/dartpy/osg/src/generated/_ZN4dart3gui3osg9WorldNodeE.cpp:1:
/usr/include/c++/5/backward/auto_ptr.h:170:21: error: within this context
       ~auto_ptr() { delete _M_ptr; }
```

If I comment out all the constructor bindings in `_ZN4dart3gui3osg9WorldNodeE.cpp`, the errors go away.

Is there a way to resolve this through `chimera` configurations? Or, does `chimera` need to skip generating constructor bindings for this case?

--- a/include/chimera/mstch.h
+++ b/include/chimera/mstch.h
@@ -68,6 +68,8 @@ public:
             });
     }
 
+    virtual ~ClangWrapper() = default;
+
     ::mstch::node last()
     {
         return last_;
@@ -211,6 +213,7 @@ public:
 
 protected:
     const std::set<const clang::CXXRecordDecl *> *available_decls_;
+    std::set<const clang::Type *> non_public_nested_decl_types_;
 };
 
 class Enum : public ClangWrapper<clang::EnumDecl>

--- a/include/chimera/mstch.h
+++ b/include/chimera/mstch.h
@@ -213,7 +213,6 @@ public:
 
 protected:
     const std::set<const clang::CXXRecordDecl *> *available_decls_;
-    std::set<const clang::Type *> non_public_nested_decl_types_;
 };
 
 class Enum : public ClangWrapper<clang::EnumDecl>

--- a/include/chimera/util.h
+++ b/include/chimera/util.h
@@ -238,6 +238,14 @@ bool needsReturnValuePolicy(const clang::NamedDecl *decl,
 std::pair<unsigned, unsigned> getFunctionArgumentRange(
     const clang::FunctionDecl *decl);
 
+/**
+ * Returns whether a method has a non-public type of parameter.
+ *
+ * This is possible when the class wants to hide the constructor from being
+ * called publicly.
+ */
+bool hasNonPublicParam(const clang::CXXMethodDecl *decl);
+
 } // namespace util
 } // namespace chimera
 

--- a/src/mstch.cpp
+++ b/src/mstch.cpp
@@ -233,6 +233,31 @@ CXXRecord::CXXRecord(const ::chimera::CompiledConfiguration &config,
             {"static_methods?",
              &CXXRecord::isNonFalse<CXXRecord, &CXXRecord::methods>},
         });
+
+    // Collect non-public nested classes
+    for (auto d = decl_->decls_begin(); d != decl_->decls_end(); ++d)
+    {
+        if (CXXRecordDecl *C = dyn_cast<CXXRecordDecl>(*d))
+        {
+            if (C->getAccess() == AS_private)
+            {
+                std::cout << "C.name (private): " << C->getNameAsString()
+                          << std::endl;
+                non_public_nested_decl_types_.insert(C->getTypeForDecl());
+            }
+            else if (C->getAccess() == AS_protected)
+            {
+                std::cout << "C.name (protected): " << C->getNameAsString()
+                          << std::endl;
+                non_public_nested_decl_types_.insert(C->getTypeForDecl());
+            }
+            else
+            {
+                std::cout << "C.name (public): " << C->getNameAsString()
+                          << std::endl;
+            }
+        }
+    }
 }
 
 ::mstch::node CXXRecord::bases()

--- a/src/mstch.cpp
+++ b/src/mstch.cpp
@@ -415,6 +415,8 @@ CXXRecord::CXXRecord(const ::chimera::CompiledConfiguration &config,
             continue;
         if (method_decl->getAccess() != AS_public)
             continue; // skip protected and private members
+        if (chimera::util::hasNonPublicParam(method_decl))
+            continue;
         if (isa<CXXConversionDecl>(method_decl))
             continue;
         if (isa<CXXDestructorDecl>(method_decl))

--- a/src/mstch.cpp
+++ b/src/mstch.cpp
@@ -233,31 +233,6 @@ CXXRecord::CXXRecord(const ::chimera::CompiledConfiguration &config,
             {"static_methods?",
              &CXXRecord::isNonFalse<CXXRecord, &CXXRecord::methods>},
         });
-
-    // Collect non-public nested classes
-    for (auto d = decl_->decls_begin(); d != decl_->decls_end(); ++d)
-    {
-        if (CXXRecordDecl *C = dyn_cast<CXXRecordDecl>(*d))
-        {
-            if (C->getAccess() == AS_private)
-            {
-                std::cout << "C.name (private): " << C->getNameAsString()
-                          << std::endl;
-                non_public_nested_decl_types_.insert(C->getTypeForDecl());
-            }
-            else if (C->getAccess() == AS_protected)
-            {
-                std::cout << "C.name (protected): " << C->getNameAsString()
-                          << std::endl;
-                non_public_nested_decl_types_.insert(C->getTypeForDecl());
-            }
-            else
-            {
-                std::cout << "C.name (public): " << C->getNameAsString()
-                          << std::endl;
-            }
-        }
-    }
 }
 
 ::mstch::node CXXRecord::bases()
@@ -385,6 +360,8 @@ CXXRecord::CXXRecord(const ::chimera::CompiledConfiguration &config,
             continue;
         if (method_decl->getAccess() != AS_public)
             continue; // skip protected and private members
+        if (chimera::util::hasNonPublicParam(method_decl))
+            continue;
         if (method_decl->isDeleted())
             continue;
         if (!isa<CXXConstructorDecl>(method_decl))

--- a/src/util.cpp
+++ b/src/util.cpp
@@ -10,6 +10,11 @@
 #include <clang/Sema/SemaDiagnostic.h>
 #include "clang/AST/DeclTemplate.h"
 
+namespace chimera
+{
+namespace util
+{
+
 using namespace clang;
 
 namespace
@@ -37,8 +42,7 @@ std::string generateUniqueName()
 
 } // namespace
 
-::mstch::node chimera::util::wrapYAMLNode(const YAML::Node &node,
-                                          chimera::util::ScalarConversionFn fn)
+::mstch::node wrapYAMLNode(const YAML::Node &node, ScalarConversionFn fn)
 {
     switch (node.Type())
     {
@@ -72,9 +76,8 @@ std::string generateUniqueName()
     }
 }
 
-void chimera::util::extendWithYAMLNode(::mstch::map &map,
-                                       const YAML::Node &node, bool overwrite,
-                                       chimera::util::ScalarConversionFn fn)
+void extendWithYAMLNode(::mstch::map &map, const YAML::Node &node,
+                        bool overwrite, ScalarConversionFn fn)
 {
     // Ignore invalid node.
     if (!node)
@@ -98,13 +101,13 @@ void chimera::util::extendWithYAMLNode(::mstch::map &map,
         auto result = map.emplace(std::make_pair(name, ::mstch::node{}));
         if (result.second || overwrite)
         {
-            result.first->second = chimera::util::wrapYAMLNode(value, fn);
+            result.first->second = wrapYAMLNode(value, fn);
         }
     }
 }
 
-const NamedDecl *chimera::util::resolveDeclaration(
-    CompilerInstance *ci, const llvm::StringRef declStr)
+const NamedDecl *resolveDeclaration(CompilerInstance *ci,
+                                    const llvm::StringRef declStr)
 {
     // Immediately return if passed an empty string.
     if (declStr.empty())
@@ -152,10 +155,9 @@ const NamedDecl *chimera::util::resolveDeclaration(
     return nullptr;
 }
 
-const QualType chimera::util::resolveType(CompilerInstance *ci,
-                                          const llvm::StringRef typeStr)
+const QualType resolveType(CompilerInstance *ci, const llvm::StringRef typeStr)
 {
-    auto decl = chimera::util::resolveDeclaration(
+    auto decl = resolveDeclaration(
         ci, "typedef " + typeStr.str() + " " + generateUniqueName());
     if (!decl)
         return emptyType_;
@@ -171,10 +173,10 @@ const QualType chimera::util::resolveType(CompilerInstance *ci,
     return typedef_decl->getUnderlyingType().getCanonicalType();
 }
 
-const RecordDecl *chimera::util::resolveRecord(CompilerInstance *ci,
-                                               const llvm::StringRef recordStr)
+const RecordDecl *resolveRecord(CompilerInstance *ci,
+                                const llvm::StringRef recordStr)
 {
-    auto type = chimera::util::resolveType(ci, recordStr).getTypePtrOrNull();
+    auto type = resolveType(ci, recordStr).getTypePtrOrNull();
     if (!type)
         return nullptr;
 
@@ -185,10 +187,10 @@ const RecordDecl *chimera::util::resolveRecord(CompilerInstance *ci,
     return cast<RecordDecl>(cxx_record_type->getCanonicalDecl());
 }
 
-const NamespaceDecl *chimera::util::resolveNamespace(
-    CompilerInstance *ci, const llvm::StringRef nsStr)
+const NamespaceDecl *resolveNamespace(CompilerInstance *ci,
+                                      const llvm::StringRef nsStr)
 {
-    auto decl = chimera::util::resolveDeclaration(
+    auto decl = resolveDeclaration(
         ci, "namespace " + generateUniqueName() + " = " + nsStr.str());
     if (!decl)
         return nullptr;
@@ -203,7 +205,7 @@ const NamespaceDecl *chimera::util::resolveNamespace(
     return cast<NamespaceAliasDecl>(decl)->getNamespace()->getCanonicalDecl();
 }
 
-std::string chimera::util::constructMangledName(const NamedDecl *decl)
+std::string constructMangledName(const NamedDecl *decl)
 {
     std::string mangled_name;
     llvm::raw_string_ostream mangled_name_stream(mangled_name);
@@ -212,7 +214,7 @@ std::string chimera::util::constructMangledName(const NamedDecl *decl)
     return mangled_name_stream.str();
 }
 
-std::string chimera::util::constructBindingName(const CXXRecordDecl *decl)
+std::string constructBindingName(const CXXRecordDecl *decl)
 {
     // If this is an anonymous struct, then use the name of its typedef.
     if (TypedefNameDecl *typedef_decl = decl->getTypedefNameForAnonDecl())
@@ -224,12 +226,12 @@ std::string chimera::util::constructBindingName(const CXXRecordDecl *decl)
 
     // If the class is a template, use the mangled string name so that it does
     // not collide with other template instantiations.
-    std::string mangled_name = chimera::util::constructMangledName(decl);
+    std::string mangled_name = constructMangledName(decl);
 
     // Throw a warning that this class name was mangled, because users will
     // probably want to override these names with more sensible ones.
     std::cerr << "Warning: The class '"
-              << chimera::util::getFullyQualifiedDeclTypeAsString(decl) << "'"
+              << getFullyQualifiedDeclTypeAsString(decl) << "'"
               << " was bound to the mangled name "
               << "'" << mangled_name << "'"
               << " because the unqualified class name of "
@@ -239,25 +241,23 @@ std::string chimera::util::constructBindingName(const CXXRecordDecl *decl)
     return mangled_name;
 }
 
-QualType chimera::util::getFullyQualifiedType(ASTContext &context, QualType qt)
+QualType getFullyQualifiedType(ASTContext &context, QualType qt)
 {
     return cling::utils::TypeName::GetFullyQualifiedType(qt, context);
 }
 
-std::string chimera::util::getFullyQualifiedTypeName(ASTContext &context,
-                                                     QualType qt)
+std::string getFullyQualifiedTypeName(ASTContext &context, QualType qt)
 {
     return cling::utils::TypeName::GetFullyQualifiedName(qt, context);
 }
 
-std::string chimera::util::getFullyQualifiedDeclTypeAsString(
-    const TypeDecl *decl)
+std::string getFullyQualifiedDeclTypeAsString(const TypeDecl *decl)
 {
-    return chimera::util::getFullyQualifiedTypeName(
-        decl->getASTContext(), QualType(decl->getTypeForDecl(), 0));
+    return getFullyQualifiedTypeName(decl->getASTContext(),
+                                     QualType(decl->getTypeForDecl(), 0));
 }
 
-bool chimera::util::isAssignable(const CXXRecordDecl *decl)
+bool isAssignable(const CXXRecordDecl *decl)
 {
     if (decl->isAbstract())
         return false;
@@ -272,7 +272,7 @@ bool chimera::util::isAssignable(const CXXRecordDecl *decl)
     return false;
 }
 
-bool chimera::util::isAssignable(ASTContext &context, QualType qual_type)
+bool isAssignable(ASTContext &context, QualType qual_type)
 {
     if (qual_type.isConstQualified())
         return false;
@@ -284,7 +284,7 @@ bool chimera::util::isAssignable(ASTContext &context, QualType qual_type)
         return qual_type.isTriviallyCopyableType(context);
 }
 
-bool chimera::util::isCopyable(const CXXRecordDecl *decl)
+bool isCopyable(const CXXRecordDecl *decl)
 {
     if (decl->isAbstract())
         return false;
@@ -298,7 +298,7 @@ bool chimera::util::isCopyable(const CXXRecordDecl *decl)
     return false;
 }
 
-bool chimera::util::isCopyable(ASTContext &context, QualType qual_type)
+bool isCopyable(ASTContext &context, QualType qual_type)
 {
     // TODO: Is this logic correct?
 
@@ -308,7 +308,7 @@ bool chimera::util::isCopyable(ASTContext &context, QualType qual_type)
         return qual_type.isTriviallyCopyableType(context);
 }
 
-bool chimera::util::isInsideTemplateClass(const DeclContext *decl_context)
+bool isInsideTemplateClass(const DeclContext *decl_context)
 {
     if (!decl_context->isRecord())
         return false;
@@ -328,7 +328,7 @@ bool chimera::util::isInsideTemplateClass(const DeclContext *decl_context)
         return false;
 }
 
-bool chimera::util::isVariadicFunctionTemplate(const FunctionTemplateDecl *decl)
+bool isVariadicFunctionTemplate(const FunctionTemplateDecl *decl)
 {
     // Based on SemaTemplateDeduction.cpp
     // http://clang.llvm.org/doxygen/SemaTemplateDeduction_8cpp_source.html
@@ -352,7 +352,7 @@ bool chimera::util::isVariadicFunctionTemplate(const FunctionTemplateDecl *decl)
     return true;
 }
 
-std::vector<std::string> chimera::util::getTemplateParameterStrings(
+std::vector<std::string> getTemplateParameterStrings(
     ASTContext &context, const ArrayRef<TemplateArgument> &params)
 {
     std::vector<std::string> outputs;
@@ -403,7 +403,7 @@ std::vector<std::string> chimera::util::getTemplateParameterStrings(
     return outputs;
 }
 
-std::string chimera::util::getTemplateParameterString(const FunctionDecl *decl)
+std::string getTemplateParameterString(const FunctionDecl *decl)
 {
     std::stringstream ss;
 
@@ -414,7 +414,7 @@ std::string chimera::util::getTemplateParameterString(const FunctionDecl *decl)
             = decl->getTemplateSpecializationArgs())
         {
             ss << "<";
-            const auto param_strs = chimera::util::getTemplateParameterStrings(
+            const auto param_strs = getTemplateParameterStrings(
                 decl->getASTContext(), params->asArray());
 
             for (size_t iparam = 0; iparam < param_strs.size(); ++iparam)
@@ -430,8 +430,7 @@ std::string chimera::util::getTemplateParameterString(const FunctionDecl *decl)
     return ss.str();
 }
 
-std::set<const CXXRecordDecl *> chimera::util::getBaseClassDecls(
-    const CXXRecordDecl *decl)
+std::set<const CXXRecordDecl *> getBaseClassDecls(const CXXRecordDecl *decl)
 {
     std::set<const CXXRecordDecl *> base_decls;
 
@@ -448,12 +447,11 @@ std::set<const CXXRecordDecl *> chimera::util::getBaseClassDecls(
     return base_decls;
 }
 
-std::set<const CXXRecordDecl *> chimera::util::getBaseClassDecls(
+std::set<const CXXRecordDecl *> getBaseClassDecls(
     const CXXRecordDecl *decl, std::set<const CXXRecordDecl *> available_decls)
 {
     // Get all base classes.
-    std::set<const CXXRecordDecl *> all_base_decls
-        = chimera::util::getBaseClassDecls(decl);
+    std::set<const CXXRecordDecl *> all_base_decls = getBaseClassDecls(decl);
 
     // Filter the classes by availability.
     std::set<const CXXRecordDecl *> base_decls;
@@ -475,7 +473,7 @@ std::set<const CXXRecordDecl *> chimera::util::getBaseClassDecls(
     return base_decls;
 }
 
-bool chimera::util::containsIncompleteType(Sema &sema, QualType qual_type)
+bool containsIncompleteType(Sema &sema, QualType qual_type)
 {
     const Type *type = qual_type.getTypePtr();
 
@@ -505,7 +503,7 @@ bool chimera::util::containsIncompleteType(Sema &sema, QualType qual_type)
     }
 }
 
-bool chimera::util::containsIncompleteType(Sema &sema, const FunctionDecl *decl)
+bool containsIncompleteType(Sema &sema, const FunctionDecl *decl)
 {
     for (unsigned int iparam = 0; iparam < decl->getNumParams(); ++iparam)
     {
@@ -522,8 +520,8 @@ bool chimera::util::containsIncompleteType(Sema &sema, const FunctionDecl *decl)
                           << "'";
 
             std::cerr << " has an incomplete type '"
-                      << chimera::util::getFullyQualifiedTypeName(
-                             decl->getASTContext(), param_decl->getType())
+                      << getFullyQualifiedTypeName(decl->getASTContext(),
+                                                   param_decl->getType())
                       << "'.\n";
             return true;
         }
@@ -531,7 +529,7 @@ bool chimera::util::containsIncompleteType(Sema &sema, const FunctionDecl *decl)
     return false;
 }
 
-bool chimera::util::containsRValueReference(const FunctionDecl *decl)
+bool containsRValueReference(const FunctionDecl *decl)
 {
     for (unsigned int iparam = 0; iparam < decl->getNumParams(); ++iparam)
     {
@@ -548,8 +546,8 @@ bool chimera::util::containsRValueReference(const FunctionDecl *decl)
                           << "'";
 
             std::cerr << " is an rvalue reference of type '"
-                      << chimera::util::getFullyQualifiedTypeName(
-                             decl->getASTContext(), param_decl->getType())
+                      << getFullyQualifiedTypeName(decl->getASTContext(),
+                                                   param_decl->getType())
                       << "'.\n";
 
             return true;
@@ -558,7 +556,7 @@ bool chimera::util::containsRValueReference(const FunctionDecl *decl)
     return false;
 }
 
-bool chimera::util::containsNonCopyableType(const FunctionDecl *decl)
+bool containsNonCopyableType(const FunctionDecl *decl)
 {
     for (unsigned int iparam = 0; iparam < decl->getNumParams(); ++iparam)
     {
@@ -567,8 +565,7 @@ bool chimera::util::containsNonCopyableType(const FunctionDecl *decl)
             = param_decl->getType().getTypePtr();
 
         if (!param_type->isReferenceType() && !param_type->isPointerType()
-            && !chimera::util::isCopyable(decl->getASTContext(),
-                                          param_decl->getType()))
+            && !isCopyable(decl->getASTContext(), param_decl->getType()))
         {
             std::cerr << "Warning: Skipped function '"
                       << decl->getQualifiedNameAsString() << "' because";
@@ -580,8 +577,8 @@ bool chimera::util::containsNonCopyableType(const FunctionDecl *decl)
                           << "'";
 
             std::cerr << " has a non-copyable type '"
-                      << chimera::util::getFullyQualifiedTypeName(
-                             decl->getASTContext(), param_decl->getType())
+                      << getFullyQualifiedTypeName(decl->getASTContext(),
+                                                   param_decl->getType())
                       << "'.\n";
 
             return true;
@@ -590,8 +587,7 @@ bool chimera::util::containsNonCopyableType(const FunctionDecl *decl)
     return false;
 }
 
-bool chimera::util::needsReturnValuePolicy(const NamedDecl *decl,
-                                           QualType return_type)
+bool needsReturnValuePolicy(const NamedDecl *decl, QualType return_type)
 {
     if (return_type.getTypePtr()->isReferenceType())
     {
@@ -616,7 +612,7 @@ bool chimera::util::needsReturnValuePolicy(const NamedDecl *decl,
     return false;
 }
 
-std::pair<unsigned, unsigned> chimera::util::getFunctionArgumentRange(
+std::pair<unsigned, unsigned> getFunctionArgumentRange(
     const clang::FunctionDecl *decl)
 {
     const unsigned max_arguments = decl->getNumParams();
@@ -630,3 +626,31 @@ std::pair<unsigned, unsigned> chimera::util::getFunctionArgumentRange(
 
     return std::pair<unsigned, unsigned>(min_arguments, max_arguments);
 }
+
+bool hasNonPublicParam(const CXXMethodDecl *decl)
+{
+    for (auto i = 0u; i < decl->getNumParams(); ++i)
+    {
+        const ParmVarDecl *param_decl = decl->getParamDecl(i);
+        QualType param_type = param_decl->getType();
+        if (param_type->isReferenceType())
+        {
+            param_type = param_type.getNonReferenceType();
+        }
+        // TODO: are there any other special cases than reference type?
+
+        if (auto record_type = dyn_cast<RecordType>(param_type))
+        {
+            CXXRecordDecl *cxx_record_decl
+                = cast<CXXRecordDecl>(record_type->getDecl());
+            if (cxx_record_decl->getAccess() != AS_public)
+            {
+                return true;
+            }
+        }
+    }
+    return false;
+}
+
+} // namespace util
+} // namespace chimera

--- a/test/examples/02_class/class.h
+++ b/test/examples/02_class/class.h
@@ -70,6 +70,18 @@ public:
   std::string m_name;
 };
 
+class MainClass
+{
+public:
+  MainClass() = default;
+
+  class NestedClass
+  {
+  public:
+    NestedClass() = default;
+  };
+};
+
 namespace detail {
 
 class ClassInDetail {};

--- a/test/examples/02_class/class.h
+++ b/test/examples/02_class/class.h
@@ -7,73 +7,75 @@
 namespace chimera_test {
 namespace nested_namespace {
 
-class Animal
+//class Animal
+//{
+//public:
+//  Animal();
+//  virtual ~Animal();
+//  virtual std::string type() const;
+//  virtual std::string pure_virtual_type() const = 0;
+//};
+
+//class Dog : public Animal
+//{
+//public:
+//  Dog();
+//  std::string type() const override;
+//  std::string pure_virtual_type() const override;
+
+//  static std::string static_type();
+//};
+
+//class Strong {};
+
+//class Husky : public Dog
+//{
+//public:
+//  Husky();
+//  std::string type() const override;
+//  std::string pure_virtual_type() const override;
+//};
+
+//class StrongHusky : public Dog, Strong
+//{
+//public:
+//  StrongHusky();
+//  std::string type() const override;
+//  std::string pure_virtual_type() const override;
+//};
+
+//class DefaultArguments
+//{
+//public:
+//  DefaultArguments() = default;
+//  int add(int i = 1, int j = 2) const;
+//};
+
+//class StaticFields
+//{
+//public:
+//  static std::string m_static_readwrite_type;
+//  static const std::string m_static_readonly_type;
+//  static std::string static_type();
+//};
+
+class NonPublicParamInConstructor
 {
+protected:
+    using ProtectedTypeDef = int;
+  struct ProtectedData{};
+  enum ProtectedEnum { PE_ONE };
+
 public:
-  Animal();
-  virtual ~Animal();
-  virtual std::string type() const;
-  virtual std::string pure_virtual_type() const = 0;
-};
-
-class Dog : public Animal
-{
-public:
-  Dog();
-  std::string type() const override;
-  std::string pure_virtual_type() const override;
-
-  static std::string static_type();
-};
-
-class Strong {};
-
-class Husky : public Dog
-{
-public:
-  Husky();
-  std::string type() const override;
-  std::string pure_virtual_type() const override;
-};
-
-class StrongHusky : public Dog, Strong
-{
-public:
-  StrongHusky();
-  std::string type() const override;
-  std::string pure_virtual_type() const override;
-};
-
-class DefaultArguments
-{
-public:
-  DefaultArguments() = default;
-  int add(int i = 1, int j = 2) const;
-};
-
-class StaticFields
-{
-public:
-  static std::string m_static_readwrite_type;
-  static const std::string m_static_readonly_type;
-  static std::string static_type();
-};
-
-class MainClass
-{
-public:
-  MainClass() = default;
-
-  class NestedClass
-  {
-  public:
-    NestedClass() = default;
-  };
+  NonPublicParamInConstructor(const ProtectedData&) {}
+  NonPublicParamInConstructor(ProtectedEnum) {}
+  NonPublicParamInConstructor(const std::string& name) : m_name(name) {}
+  std::string m_name;
 };
 
 namespace detail {
 
-class ClassInDetail {};
+//class ClassInDetail {};
 
 } // namespace detail
 

--- a/test/examples/02_class/class.h
+++ b/test/examples/02_class/class.h
@@ -7,75 +7,72 @@
 namespace chimera_test {
 namespace nested_namespace {
 
-//class Animal
-//{
-//public:
-//  Animal();
-//  virtual ~Animal();
-//  virtual std::string type() const;
-//  virtual std::string pure_virtual_type() const = 0;
-//};
+class Animal
+{
+public:
+ Animal();
+ virtual ~Animal();
+ virtual std::string type() const;
+ virtual std::string pure_virtual_type() const = 0;
+};
 
-//class Dog : public Animal
-//{
-//public:
-//  Dog();
-//  std::string type() const override;
-//  std::string pure_virtual_type() const override;
+class Dog : public Animal
+{
+public:
+ Dog();
+ std::string type() const override;
+ std::string pure_virtual_type() const override;
 
-//  static std::string static_type();
-//};
+ static std::string static_type();
+};
 
-//class Strong {};
+class Strong {};
 
-//class Husky : public Dog
-//{
-//public:
-//  Husky();
-//  std::string type() const override;
-//  std::string pure_virtual_type() const override;
-//};
+class Husky : public Dog
+{
+public:
+ Husky();
+ std::string type() const override;
+ std::string pure_virtual_type() const override;
+};
 
-//class StrongHusky : public Dog, Strong
-//{
-//public:
-//  StrongHusky();
-//  std::string type() const override;
-//  std::string pure_virtual_type() const override;
-//};
+class StrongHusky : public Dog, Strong
+{
+public:
+ StrongHusky();
+ std::string type() const override;
+ std::string pure_virtual_type() const override;
+};
 
-//class DefaultArguments
-//{
-//public:
-//  DefaultArguments() = default;
-//  int add(int i = 1, int j = 2) const;
-//};
+class DefaultArguments
+{
+public:
+ DefaultArguments() = default;
+ int add(int i = 1, int j = 2) const;
+};
 
-//class StaticFields
-//{
-//public:
-//  static std::string m_static_readwrite_type;
-//  static const std::string m_static_readonly_type;
-//  static std::string static_type();
-//};
+class StaticFields
+{
+public:
+ static std::string m_static_readwrite_type;
+ static const std::string m_static_readonly_type;
+ static std::string static_type();
+};
 
 class NonPublicParamInConstructor
 {
 protected:
-    using ProtectedTypeDef = int;
   struct ProtectedData{};
-  enum ProtectedEnum { PE_ONE };
 
 public:
   NonPublicParamInConstructor(const ProtectedData&) {}
-  NonPublicParamInConstructor(ProtectedEnum) {}
   NonPublicParamInConstructor(const std::string& name) : m_name(name) {}
   std::string m_name;
 };
 
 namespace detail {
 
-//class ClassInDetail {};
+class ClassInDetail {};
 
 } // namespace detail
 

--- a/test/examples/02_class/class.h
+++ b/test/examples/02_class/class.h
@@ -10,20 +10,20 @@ namespace nested_namespace {
 class Animal
 {
 public:
- Animal();
- virtual ~Animal();
- virtual std::string type() const;
- virtual std::string pure_virtual_type() const = 0;
+  Animal();
+  virtual ~Animal();
+  virtual std::string type() const;
+  virtual std::string pure_virtual_type() const = 0;
 };
 
 class Dog : public Animal
 {
 public:
- Dog();
- std::string type() const override;
- std::string pure_virtual_type() const override;
+  Dog();
+  std::string type() const override;
+  std::string pure_virtual_type() const override;
 
- static std::string static_type();
+  static std::string static_type();
 };
 
 class Strong {};
@@ -31,32 +31,32 @@ class Strong {};
 class Husky : public Dog
 {
 public:
- Husky();
- std::string type() const override;
- std::string pure_virtual_type() const override;
+  Husky();
+  std::string type() const override;
+  std::string pure_virtual_type() const override;
 };
 
 class StrongHusky : public Dog, Strong
 {
 public:
- StrongHusky();
- std::string type() const override;
- std::string pure_virtual_type() const override;
+  StrongHusky();
+  std::string type() const override;
+  std::string pure_virtual_type() const override;
 };
 
 class DefaultArguments
 {
 public:
- DefaultArguments() = default;
- int add(int i = 1, int j = 2) const;
+  DefaultArguments() = default;
+  int add(int i = 1, int j = 2) const;
 };
 
 class StaticFields
 {
 public:
- static std::string m_static_readwrite_type;
- static const std::string m_static_readonly_type;
- static std::string static_type();
+  static std::string m_static_readwrite_type;
+  static const std::string m_static_readonly_type;
+  static std::string static_type();
 };
 
 class NonPublicParamInConstructor

--- a/test/examples/02_class/class.py
+++ b/test/examples/02_class/class.py
@@ -40,6 +40,9 @@ class TestFunction(unittest.TestCase):
         self.assertEqual(py11.StaticFields.m_static_readwrite_type, 'new type')
         self.assertEqual(py11.StaticFields.static_type(), 'static type')
 
+        non_pub_param_in_ctr = py11.NonPublicParamInConstructor('Herb')
+        self.assertEqual(non_pub_param_in_ctr.m_name, 'Herb')
+
         # Check if the main class and the nested class can be created w/o errors
         mc = py11.MainClass()
         nc = py11.MainClass.NestedClass()
@@ -78,6 +81,9 @@ class TestFunction(unittest.TestCase):
         # boost.StaticFields.m_static_readwrite_type = 'new type'
         # self.assertEqual(boost.StaticFields.m_static_readwrite_type, 'new type')
         self.assertEqual(boost.StaticFields.static_type(), 'static type')
+
+        non_pub_param_in_ctr = boost.NonPublicParamInConstructor('Herb')
+        self.assertEqual(non_pub_param_in_ctr.m_name, 'Herb')
 
         # Check if the main class and the nested class can be created w/o errors
         mc = boost.MainClass()

--- a/test/test_emulator.cpp
+++ b/test/test_emulator.cpp
@@ -26,7 +26,7 @@ TEST(Emulator, GeneralOptions)
 }
 
 //==============================================================================
- TEST(Emulator, 01_Function)
+TEST(Emulator, 01_Function)
 {
     Emulator e;
     e.SetSource("01_function/function.h");

--- a/test/test_emulator.cpp
+++ b/test/test_emulator.cpp
@@ -26,14 +26,17 @@ TEST(Emulator, GeneralOptions)
 }
 
 //==============================================================================
-// TEST(Emulator, 01_Function)
-//{
-//    Emulator e;
-//    e.SetSource("01_function/function.h");
-//    e.SetConfigurationFile("01_function/function_pybind11.yaml");
-//    e.SetBinding("pybind11");
-//    EXPECT_EXIT(e.Run(), ::testing::ExitedWithCode(0), ".*");
-//}
+ TEST(Emulator, 01_Function)
+{
+    Emulator e;
+    e.SetSource("01_function/function.h");
+    e.SetConfigurationFile("01_function/function_pybind11.yaml");
+    e.SetBinding("pybind11");
+
+    // EXPECT_EXIT is necessary to continue to run subsequent tests, but it
+    // doesn't stop at the breakpoints. For debugging use e.Run() instead.
+    EXPECT_EXIT(e.Run(), ::testing::ExitedWithCode(0), ".*");
+}
 
 //==============================================================================
 TEST(Emulator, 02_Class)
@@ -42,5 +45,8 @@ TEST(Emulator, 02_Class)
     e.SetSource("02_class/class.h");
     e.SetConfigurationFile("02_class/class.yaml");
     e.SetBinding("pybind11");
+
+    // EXPECT_EXIT is necessary to continue to run subsequent tests, but it
+    // doesn't stop at the breakpoints. For debugging use e.Run() instead.
     EXPECT_EXIT(e.Run(), ::testing::ExitedWithCode(0), ".*");
 }

--- a/test/test_emulator.cpp
+++ b/test/test_emulator.cpp
@@ -26,14 +26,14 @@ TEST(Emulator, GeneralOptions)
 }
 
 //==============================================================================
-TEST(Emulator, 01_Function)
-{
-    Emulator e;
-    e.SetSource("01_function/function.h");
-    e.SetConfigurationFile("01_function/function_pybind11.yaml");
-    e.SetBinding("pybind11");
-    e.Run();
-}
+// TEST(Emulator, 01_Function)
+//{
+//    Emulator e;
+//    e.SetSource("01_function/function.h");
+//    e.SetConfigurationFile("01_function/function_pybind11.yaml");
+//    e.SetBinding("pybind11");
+//    EXPECT_EXIT(e.Run(), ::testing::ExitedWithCode(0), ".*");
+//}
 
 //==============================================================================
 TEST(Emulator, 02_Class)
@@ -42,5 +42,5 @@ TEST(Emulator, 02_Class)
     e.SetSource("02_class/class.h");
     e.SetConfigurationFile("02_class/class.yaml");
     e.SetBinding("pybind11");
-    e.Run();
+    EXPECT_EXIT(e.Run(), ::testing::ExitedWithCode(0), ".*");
 }

